app/controllers/BalanceRequestResponseController.scala

```diff
@@ -76,29 +78,52 @@ class BalanceRequestResponseController @Inject() (
     implicit request =>
       withMetricsTimerResult(UpdateBalanceRequest) {
         requireMessageTypeHeader { messageType =>
-          cache
-            .updateBalance(recipient, messageType, request.body)
-            .flatTap(logServiceError("updating balance request", _))
-            .map {
-              case Right(_) =>
-                Ok
-              case Left(error @ BadRequestError(_)) =>
-                BadRequest(Json.toJson[BalanceRequestError](error))
-              case Left(error @ XmlValidationError(_, _)) =>
-                BadRequest(Json.toJson[BalanceRequestError](error))
-              case Left(error @ NotFoundError(_)) =>
-                NotFound(Json.toJson[BalanceRequestError](error))
-              case Left(_) =>
-                InternalServerError(Json.toJson(BalanceRequestError.internalServiceError()))
-            }
-            .recoverWith { case NonFatal(e) =>
-              logger.error(e)("Unhandled exception thrown").map { _ =>
-                val error     = InternalServiceError.causedBy(e)
-                val errorJson = Json.toJson[BalanceRequestError](error)
-                InternalServerError(errorJson)
-              }
-            }
+          for {
+            result <- updateBalanceRequestIO(recipient, messageType)
+            _      <- recordRequestLength(request)
+          } yield result
         }
       }
   }
+
+  private def updateBalanceRequestIO(
```

Comment:
I don't have a great suggestion here but I don't really love the name of this function - perhaps `updateBalanceRequestRecord` or something like that? Or perhaps the `recordRequestLength` action could become a `flatTap` side-effect on the original expression?